algorithms/src/snark/marlin/ahp/prover/oracles.rs

```diff
@@ -18,33 +18,47 @@ use std::collections::BTreeMap;
 
 use snarkvm_fields::PrimeField;
 
-use crate::polycommit::sonic_pc::{LabeledPolynomial, LabeledPolynomialWithBasis, PolynomialInfo, PolynomialLabel};
+use crate::{
+    polycommit::sonic_pc::{LabeledPolynomial, LabeledPolynomialWithBasis, PolynomialInfo, PolynomialLabel},
+    snark::marlin::{ahp::indexer::Circuit, CircuitId, MarlinMode},
+};
 
 /// The first set of prover oracles.
 #[derive(Debug, Clone)]
-pub struct FirstOracles<F: PrimeField> {
-    pub(in crate::snark::marlin) batches: Vec<SingleEntry<F>>,
+pub struct FirstOracles<'a, F: PrimeField, MM: MarlinMode> {
+    pub(in crate::snark::marlin) batches: BTreeMap<&'a Circuit<F, MM>, Vec<SingleEntry<F>>>,
```

Comment:
Can we get away with just putting a `CircuitId` here, instead of an `&'a Circuit`?

Reply:
Done!

Reply:
Hm I think this is still a `Circuit`, no? What I mean is that we should change the type of `batches` to `BTreeMap<CircuitId, ...>`.

Reply:
The source Github shows is outdated, it is CircuitId now